common/lib/transport/connectionmanager.js

```diff
@@ -809,70 +785,217 @@ var ConnectionManager = (function() {
+		var self = this,
+			giveUp = function(err) {
 				self.notifyState({state: self.states.connecting.failState, error: err});
-		};
+			},
+			candidateHosts = this.httpHosts.slice(),
+			hostAttemptCb = function(err) {
+				if(err) {
+					var wasFatal = self.failConnectionIfFatal(err);
+					if(!wasFatal) {
+						tryFallbackHosts();
+						return;
+					}
+				}
+			};
 
-		var tryConnect = function() {
-			self.chooseTransport(function(err) {
+		Logger.logAction(Logger.LOG_MINOR, 'ConnectionManager.connectBase()', 'Trying to connect with base transport ' + this.baseTransport);
+
+		/* first try to establish a connection with the priority host with http transport */
+		var host = candidateHosts.shift();
+		if(!host) {
+			giveUp(new ErrorInfo('Unable to connect (no available host)', 80000, 404));
+			return;
+		}
+		transportParams.host = host;
+
+		/* this is what we'll be doing if the attempt for the main host fails */
+		function tryFallbackHosts() {
+			/* if there aren't any fallback hosts, fail */
+			if(!candidateHosts.length) {
+				giveUp(new ErrorInfo('Unable to connect (and no more fallback hosts to try)', 80000, 404));
+				return;
+			}
+			/* before trying any fallback (or any remaining fallback) we decide if
+			 * there is a problem with the ably host, or there is a general connectivity
+			 * problem */
+			var connectivityCheckTransport = self.baseTransport === 'web_socket' ? 'xhr_polling' : self.baseTransport;
+			ConnectionManager.supportedTransports[connectivityCheckTransport].checkConnectivity(function(err, connectivity) {
+				/* we know err won't happen but handle it here anyway */
 				if(err) {
-					connectErr(err);
+					giveUp(err);
 					return;
 				}
-				/* nothing to do .. as transport connection is initiated
-				 * in chooseTransport() */
+				if(!connectivity) {
+					/* the internet isn't reachable, so don't try the fallback hosts */
+					giveUp(new ErrorInfo('Unable to connect (network unreachable)', 80000, 404));
+					return;
+				}
+				/* the network is there, so there's a problem with the main host, or
+				 * its dns. Try the fallback hosts. We could try them simultaneously but
+				 * that would potentially cause a huge spike in load on the load balancer */
+				transportParams.host = Utils.arrPopRandomElement(candidateHosts);
+				self.tryATransport(transportParams, self.baseTransport, hostAttemptCb);
 			});
-		};
+		}
 
-		if(auth.method == 'basic') {
-			tryConnect();
-		} else {
-			var authOptions = (this.errorReason && Auth.isTokenErr(this.errorReason)) ? {force: true} : null;
-			auth.authorise(null, authOptions, function(err) {
-				if(err)
-					connectErr(err);
-				else
-					tryConnect();
-			});
+		this.tryATransport(transportParams, this.baseTransport, hostAttemptCb);
+	};
+
+
+	ConnectionManager.prototype.getUpgradePossibilities = function() {
+		var current = this.activeProtocol.getTransport().shortName;
+
+		if(current === 'web_socket') {
```

Comment:
I'm wondering if there's a way for this part of the logic to be less hard-coded.